crates/lox_core/src/ephemeris/daf_spk/api.rs

```diff
@@ -0,0 +1,125 @@
+        &'a self,
+        array: &'a SpkType2Array,
+        initial_epoch: Epoch,
+        epoch: Epoch,
+    ) -> Result<(&SpkType2Record, f64), DafSpkError> {
+        let seconds_from_record_start = epoch - initial_epoch;
+
+        let intlen = array.intlen as f64;
+        let mut record_number = (seconds_from_record_start / intlen).floor() as usize;
+        let mut fraction = seconds_from_record_start % intlen;
+
+        // @TODO clarify this
+        if record_number == array.n as usize {
+            record_number -= 1;
+            fraction = array.intlen as f64;
+        }
+
+        let record = array
+            .records
+            .get(record_number)
+            .ok_or(DafSpkError::UnableToFindMatchingRecord)?;
+
+        Ok((record, fraction))
+    }
+
+    pub fn position(
+        &self,
+        epoch: Epoch,
+        origin: Body,
+        target: Body,
+    ) -> Result<Position, DafSpkError> {
+        let (segment, sign) = self.find_segment(origin, target)?;
+        let sign = sign as f64;
+
+        if epoch < segment.initial_epoch || epoch > segment.final_epoch {
+            return Err(DafSpkError::UnableToFindMatchingSegment);
+        }
+
+        // @TODO use a more flexible way
+        let mut x = 0f64;
+        let mut y = 0f64;
+        let mut z = 0f64;
+
+        match &segment.data {
+            super::parser::SpkArray::Type2(array) => {
+                let (record, fraction) = self.find_record(array, segment.initial_epoch, epoch)?;
+
+                let degree_of_polynomial = array.degree_of_polynomial() as usize;
+                // @TODO figure name
+                let mut figure_out_name_for_this_var =
+                    Vec::<f64>::with_capacity(degree_of_polynomial);
+
+                figure_out_name_for_this_var.push(1f64);
+                figure_out_name_for_this_var.push(2f64 * fraction / array.intlen as f64 - 1f64);
+
+                for i in 2..degree_of_polynomial {
+                    figure_out_name_for_this_var.push(
+                        2f64 * figure_out_name_for_this_var[1]
+                            * figure_out_name_for_this_var[i - 1]
+                            - figure_out_name_for_this_var[i - 2],
+                    );
+                }
+
+                #[allow(clippy::needless_range_loop)]
+                for i in 0..degree_of_polynomial {
+                    x += sign * record.x[i] * figure_out_name_for_this_var[i];
+                    y += sign * record.y[i] * figure_out_name_for_this_var[i];
+                    z += sign * record.z[i] * figure_out_name_for_this_var[i];
+                }
+            }
+        }
+
+        Ok((x, y, z))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::ephemeris::daf_spk::parser::parse_daf_spk;
+    use crate::ephemeris::daf_spk::parser::test::FILE_CONTENTS;
+
+    use super::*;
+
+    #[test]
+    fn test_position() {
+        let spk = parse_daf_spk(&FILE_CONTENTS).expect("Unable to parse DAF/SPK");
```

Comment:
Will we always be using one standard SPK file (I'm not an astrodynamicist 😅), or are there many possible files that might need to be parsed? If so, it would be good to create some pathological SPK files to exercise the edge cases.

Reply:
I guess @helgee can give some input. To me it seems that the files themselves are sort of standard references in the industry. ESA's calceph library seems to validate against a bunch of files directly downloaded from an FTp server. 

Other libraries don't seem to be too bothered by extensive testing. The biggest limitation seems to be file size, if we synthesize some files, we need to do it in a way which doesn't bloat the repository. I'm open to suggestions.

Reply:
We should be able to build files that are pathological in one specific case quite compactly, but if there are relatively few standard files, then it's probably not worth it, unless we were to split the parsers into their own libraries at some point. Not one for now, anyway.

Reply:
We should be using SPK files as little as possible and the ones we are going to use come straight from NASA or ESA and will definitely be valid. The only issue I can think of would be files that been damaged during transfer but then it's the responsibility of the user to validate them against checksums.

My midterm goal would be to develop a more cache- and parallelisation-friendly format for ephemerides (possibly based on [rkyv](https://github.com/rkyv/rkyv) or an open standard like Apache Arrow) and provide a tool to convert SPK files to this format.

Reply:
So I think it's a good focus of effort to not go too deep in this regard.